random failpoint test faild for tiflash crash

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```
[2022/11/23 10:10:42.938 +08:00] [ERROR] [BaseDaemon.cpp:377] [########################################] [source=BaseDaemon] [thread_id=2125]
[2022/11/23 10:10:42.951 +08:00] [ERROR] [BaseDaemon.cpp:378] ["(from thread 2056) Received signal Aborted(6)."] [source=BaseDaemon] [thread_id=2125]
[2022/11/23 10:10:43.117 +08:00] [ERROR] [BaseDaemon.cpp:570] ["\n       0x9240fbd\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+153358269]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:221\n  0x7fe3c3674630\t<unknown symbol> [libpthread.so.0+63024]\n  0x7fe3c29a9387\tgsignal [libc.so.6+222087]\n  0x7fe3c29aaa78\t__GI_abort [libc.so.6+227960]\n  0x7fe3c29a21a6\t__assert_fail_base [libc.so.6+192934]\n  0x7fe3c29a2252\t__assert_fail [libc.so.6+193106]\n       0x3985972\tDB::StreamingDAGResponseWriter<std::__1::shared_ptr<DB::MPPTunnelSet> >::finishWrite() [tiflash+60316018]\n                \tdbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.cpp:66\n       0xa347922\tDB::ExchangeSenderBlockInputStream::readSuffixImpl() [tiflash+171211042]\n                \tdbms/src/DataStreams/ExchangeSenderBlockInputStream.h:49\n       0x926d762\tDB::IProfilingBlockInputStream::readSuffix() [tiflash+153540450]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:141\n       0xa34b8ce\tDB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::readSuffix() [tiflash+171227342]\n                \tdbms/src/DataStreams/UnionBlockInputStream.h:255\n       0xa394724\tDB::MPPTask::runImpl() [tiflash+171525924]\n                \tdbms/src/Flash/Mpp/MPPTask.cpp:382\n       0xa39834d\tDB::MPPTask::run()::$_12::operator()() const [tiflash+171541325]\n                \tdbms/src/Flash/Mpp/MPPTask.cpp:117\n       0xa398315\tdecltype(static_cast<DB::MPPTask::run()::$_12&>(fp)()) std::__1::__invoke<DB::MPPTask::run()::$_12&>(DB::MPPTask::run()::$_12&) [tiflash+171541269]\n                \t/usr/local/bin/../include/c++/v1/type_traits:3918\n       0xa3982cd\tvoid std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::MPPTask::run()::$_12&>(DB::MPPTask::run()::$_12&) [tiflash+171541197]\n                \t/usr/local/bin/../include/c++/v1/__functional/invoke.h:61\n       0xa39829d\tstd::__1::__function::__alloc_func<DB::MPPTask::run()::$_12, std::__1::allocator<DB::MPPTask::run()::$_12>, void ()>::operator()() [tiflash+171541149]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:171\n       0xa3973c9\tstd::__1::__function::__func<DB::MPPTask::run()::$_12, std::__1::allocator<DB::MPPTask::run()::$_12>, void ()>::operator()() [tiflash+171537353]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:345\n       0x3b79246\tstd::__1::__function::__value_func<void ()>::operator()() const [tiflash+62362182]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:498\n       0x3b791f5\tstd::__1::function<void ()>::operator()() const [tiflash+62362101]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:1175\n       0x3b85c75\tdecltype(static_cast<std::__1::function<void ()>>(fp)()) std::__1::__invoke_constexpr<std::__1::function<void ()> >(std::__1::function<void ()>&&) [tiflash+62413941]\n                \t/usr/local/bin/../include/c++/v1/type_traits:3924\n       0x3b85c51\tdecltype(auto) std::__1::__apply_tuple_impl<std::__1::function<void ()>, std::__1::tuple<> >(std::__1::function<void ()>&&, std::__1::tuple<>&&, std::__1::__tuple_indices<>) [tiflash+62413905]\n                \t/usr/local/bin/../include/c++/v1/tuple:1536\n       0x3b85c21\tdecltype(auto) std::__1::apply<std::__1::function<void ()>, std::__1::tuple<> >(std::__1::function<void ()>&&, std::__1::tuple<>&&) [tiflash+62413857]\n                \t/usr/local/bin/../include/c++/v1/tuple:1545\n       0x3b85bbc\tauto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()::operator()() [tiflash+62413756]\n                \tdbms/src/Common/wrapInvocable.h:36\n       0x3b866c9\tDB::ExecutableTask<auto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()>::execute() [tiflash+62416585]\n                \tdbms/src/Common/ExecutableTask.h:52\n       0x3b8b15d\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+62435677]\n                \tdbms/src/Common/DynamicThreadPool.cpp:101\n       0x3b8b023\tDB::DynamicThreadPool::dynamicWork(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >) [tiflash+62435363]\n                \tdbms/src/Common/DynamicThreadPool.cpp:142\n       0x3b8f750\tdecltype(*(static_cast<DB::DynamicThreadPool*>(fp0)).*fp(static_cast<std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >>(fp1))) std::__1::__invoke<void (DB::DynamicThreadPool::* const&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >, void>(void (DB::DynamicThreadPool::* const&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&) [tiflash+62453584]\n                \t/usr/local/bin/../include/c++/v1/type_traits:3859\n       0x3b8f676\tstd::__1::invoke_result<void (DB::DynamicThreadPool::* const&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >::type std::__1::invoke<void (DB::DynamicThreadPool::* const&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(void (DB::DynamicThreadPool::* const&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&) [tiflash+62453366]\n                \t/usr/local/bin/../include/c++/v1/__functional/invoke.h:93\n       0x3b8f5b5\tauto std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...)::operator()<DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(auto&&...) const [tiflash+62453173]\n                \tdbms/src/Common/ThreadFactory.h:47\n       0x3b8f3f5\tdecltype(static_cast<void (DB::DynamicThreadPool::*>(fp)(static_cast<DB::DynamicThreadPool*>(fp0), static_cast<std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >>(fp0))) std::__1::__invoke<std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&) [tiflash+62452725]\n                \t/usr/local/bin/../include/c++/v1/type_traits:3918\n       0x3b8f3c6\tvoid std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >, 2ul, 3ul>(std::__1::tuple<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >&, std::__1::__tuple_indices<2ul, 3ul>) [tiflash+62452678]\n                \t/usr/local/bin/../include/c++/v1/thread:280\n       0x3b8f05b\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > > >(void*) [tiflash+62451803]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7fe3c366cea5\tstart_thread [libpthread.so.0+32421]"] [source=BaseDaemon] [thread_id=2125]
```
### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiFlash version? (Required)
master

--- a/dbms/src/DataStreams/ExchangeSenderBlockInputStream.h
+++ b/dbms/src/DataStreams/ExchangeSenderBlockInputStream.h
@@ -46,7 +46,6 @@ protected:
     }
     void readSuffixImpl() override
     {
-        writer->finishWrite();
         LOG_DEBUG(log, "finish write with {} rows", total_rows);
     }
 

--- a/dbms/src/Flash/Coprocessor/DAGBlockOutputStream.cpp
+++ b/dbms/src/Flash/Coprocessor/DAGBlockOutputStream.cpp
@@ -35,7 +35,6 @@ void DAGBlockOutputStream::writeSuffix()
 {
     // todo error handle
     response_writer->flush();
-    response_writer->finishWrite();
 }
 
 } // namespace DB

--- a/dbms/src/Flash/Coprocessor/DAGDriver.cpp
+++ b/dbms/src/Flash/Coprocessor/DAGDriver.cpp
@@ -110,6 +110,11 @@ try
             dag_context);
         dag_output_stream = std::make_shared<DAGBlockOutputStream>(streams.in->getHeader(), std::move(response_writer));
         copyData(*streams.in, *dag_output_stream);
+        if (dag_context.collect_execution_summaries)
+        {
+            ExecutionSummaryCollector summary_collector(dag_context);
+            summary_collector.addExecuteSummaries(*dag_response);
+        }
     }
     else
     {

--- a/dbms/src/Flash/Coprocessor/DAGResponseWriter.h
+++ b/dbms/src/Flash/Coprocessor/DAGResponseWriter.h
@@ -30,7 +30,6 @@ public:
     virtual void write(const Block & block) = 0;
     /// flush cached blocks for batch writer
     virtual void flush() = 0;
-    virtual void finishWrite() = 0;
     virtual ~DAGResponseWriter() = default;
     const DAGContext & dagContext() const { return dag_context; }
 

--- a/dbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.cpp
+++ b/dbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.cpp
@@ -61,12 +61,6 @@ StreamingDAGResponseWriter<StreamWriterPtr>::StreamingDAGResponseWriter(
 }
 
 template <class StreamWriterPtr>
-void StreamingDAGResponseWriter<StreamWriterPtr>::finishWrite()
-{
-    assert(0 == rows_in_blocks);
-}
-
-template <class StreamWriterPtr>
 void StreamingDAGResponseWriter<StreamWriterPtr>::flush()
 {
     if (rows_in_blocks > 0)
@@ -80,9 +74,9 @@ void StreamingDAGResponseWriter<StreamWriterPtr>::write(const Block & block)
         block.columns() == dag_context.result_field_types.size(),
         "Output column size mismatch with field type size");
     size_t rows = block.rows();
-    rows_in_blocks += rows;
     if (rows > 0)
     {
+        rows_in_blocks += rows;
         blocks.push_back(block);
     }
 

--- a/dbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.h
+++ b/dbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.h
@@ -37,7 +37,6 @@ public:
         DAGContext & dag_context_);
     void write(const Block & block) override;
     void flush() override;
-    void finishWrite() override;
 
 private:
     void encodeThenWriteBlocks();

--- a/dbms/src/Flash/Coprocessor/UnaryDAGResponseWriter.cpp
+++ b/dbms/src/Flash/Coprocessor/UnaryDAGResponseWriter.cpp
@@ -71,19 +71,14 @@ void UnaryDAGResponseWriter::appendWarningsToDAGResponse()
     dag_response->set_warning_count(dag_context.getWarningCount());
 }
 
-void UnaryDAGResponseWriter::finishWrite()
+void UnaryDAGResponseWriter::flush()
 {
     if (current_records_num > 0)
     {
         encodeChunkToDAGResponse();
     }
+    // TODO separate from UnaryDAGResponseWriter and support mpp/batchCop.
     appendWarningsToDAGResponse();
-
-    if (dag_context.collect_execution_summaries)
-    {
-        ExecutionSummaryCollector summary_collector(dag_context);
-        summary_collector.addExecuteSummaries(*dag_response);
-    }
 }
 
 void UnaryDAGResponseWriter::write(const Block & block)

--- a/dbms/src/Flash/Coprocessor/UnaryDAGResponseWriter.h
+++ b/dbms/src/Flash/Coprocessor/UnaryDAGResponseWriter.h
@@ -38,8 +38,7 @@ public:
         DAGContext & dag_context_);
 
     void write(const Block & block) override;
-    void flush() override {}
-    void finishWrite() override;
+    void flush() override;
     void encodeChunkToDAGResponse();
     void appendWarningsToDAGResponse();
 

--- a/dbms/src/Flash/Coprocessor/tests/gtest_streaming_writer.cpp
+++ b/dbms/src/Flash/Coprocessor/tests/gtest_streaming_writer.cpp
@@ -145,7 +145,6 @@ try
         for (const auto & block : blocks)
             dag_writer->write(block);
         dag_writer->flush();
-        dag_writer->finishWrite();
 
         // 4. Start to check write_report.
         size_t expect_rows = block_rows * block_num;

--- a/dbms/src/Flash/Coprocessor/tests/gtest_ti_remote_block_inputstream.cpp
+++ b/dbms/src/Flash/Coprocessor/tests/gtest_ti_remote_block_inputstream.cpp
@@ -306,7 +306,6 @@ public:
         for (const auto & block : source_blocks)
             dag_writer->write(block);
         dag_writer->flush();
-        dag_writer->finishWrite();
 
         // 3. send execution summary
         writer->add_summary = true;
@@ -333,7 +332,6 @@ public:
         for (const auto & block : source_blocks)
             dag_writer->write(block);
         dag_writer->flush();
-        dag_writer->finishWrite();
 
         // 3. send execution summary
         writer->add_summary = true;

--- a/dbms/src/Flash/Mpp/BroadcastOrPassThroughWriter.cpp
+++ b/dbms/src/Flash/Mpp/BroadcastOrPassThroughWriter.cpp
@@ -34,12 +34,6 @@ BroadcastOrPassThroughWriter<ExchangeWriterPtr>::BroadcastOrPassThroughWriter(
 }
 
 template <class ExchangeWriterPtr>
-void BroadcastOrPassThroughWriter<ExchangeWriterPtr>::finishWrite()
-{
-    assert(0 == rows_in_blocks);
-}
-
-template <class ExchangeWriterPtr>
 void BroadcastOrPassThroughWriter<ExchangeWriterPtr>::flush()
 {
     if (rows_in_blocks > 0)
@@ -53,9 +47,9 @@ void BroadcastOrPassThroughWriter<ExchangeWriterPtr>::write(const Block & block)
         block.columns() == dag_context.result_field_types.size(),
         "Output column size mismatch with field type size");
     size_t rows = block.rows();
-    rows_in_blocks += rows;
     if (rows > 0)
     {
+        rows_in_blocks += rows;
         blocks.push_back(block);
     }
 

--- a/dbms/src/Flash/Mpp/BroadcastOrPassThroughWriter.h
+++ b/dbms/src/Flash/Mpp/BroadcastOrPassThroughWriter.h
@@ -32,7 +32,6 @@ public:
         DAGContext & dag_context_);
     void write(const Block & block) override;
     void flush() override;
-    void finishWrite() override;
 
 private:
     void encodeThenWriteBlocks();

--- a/dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp
+++ b/dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp
@@ -45,12 +45,6 @@ FineGrainedShuffleWriter<ExchangeWriterPtr>::FineGrainedShuffleWriter(
 }
 
 template <class ExchangeWriterPtr>
-void FineGrainedShuffleWriter<ExchangeWriterPtr>::finishWrite()
-{
-    assert(0 == rows_in_blocks);
-}
-
-template <class ExchangeWriterPtr>
 void FineGrainedShuffleWriter<ExchangeWriterPtr>::prepare(const Block & sample_block)
 {
     /// Initialize header block, use column type to create new empty column to handle potential null column cases
@@ -85,9 +79,9 @@ void FineGrainedShuffleWriter<ExchangeWriterPtr>::write(const Block & block)
         "Output column size mismatch with field type size");
 
     size_t rows = block.rows();
-    rows_in_blocks += rows;
     if (rows > 0)
     {
+        rows_in_blocks += rows;
         blocks.push_back(block);
     }
 

--- a/dbms/src/Flash/Mpp/FineGrainedShuffleWriter.h
+++ b/dbms/src/Flash/Mpp/FineGrainedShuffleWriter.h
@@ -36,7 +36,6 @@ public:
     void prepare(const Block & sample_block) override;
     void write(const Block & block) override;
     void flush() override;
-    void finishWrite() override;
 
 private:
     void batchWriteFineGrainedShuffle();

--- a/dbms/src/Flash/Mpp/HashPartitionWriter.cpp
+++ b/dbms/src/Flash/Mpp/HashPartitionWriter.cpp
@@ -41,12 +41,6 @@ HashPartitionWriter<ExchangeWriterPtr>::HashPartitionWriter(
 }
 
 template <class ExchangeWriterPtr>
-void HashPartitionWriter<ExchangeWriterPtr>::finishWrite()
-{
-    assert(0 == rows_in_blocks);
-}
-
-template <class ExchangeWriterPtr>
 void HashPartitionWriter<ExchangeWriterPtr>::flush()
 {
     if (rows_in_blocks > 0)
@@ -60,9 +54,9 @@ void HashPartitionWriter<ExchangeWriterPtr>::write(const Block & block)
         block.columns() == dag_context.result_field_types.size(),
         "Output column size mismatch with field type size");
     size_t rows = block.rows();
-    rows_in_blocks += rows;
     if (rows > 0)
     {
+        rows_in_blocks += rows;
         blocks.push_back(block);
     }
 

--- a/dbms/src/Flash/Mpp/HashPartitionWriter.h
+++ b/dbms/src/Flash/Mpp/HashPartitionWriter.h
@@ -34,7 +34,6 @@ public:
         DAGContext & dag_context_);
     void write(const Block & block) override;
     void flush() override;
-    void finishWrite() override;
 
 private:
     void partitionAndEncodeThenWriteBlocks();

--- a/dbms/src/Flash/Mpp/tests/gtest_mpp_exchange_writer.cpp
+++ b/dbms/src/Flash/Mpp/tests/gtest_mpp_exchange_writer.cpp
@@ -172,7 +172,6 @@ try
     dag_writer->prepare(block.cloneEmpty());
     dag_writer->write(block);
     dag_writer->flush();
-    dag_writer->finishWrite();
 
     // 4. Start to check write_report.
     std::vector<Block> decoded_blocks;
@@ -233,7 +232,6 @@ try
     for (const auto & block : blocks)
         dag_writer->write(block);
     dag_writer->flush();
-    dag_writer->finishWrite();
 
     // 4. Start to check write_report.
     size_t per_part_rows = block_rows * block_num / part_num;
@@ -294,7 +292,6 @@ try
     for (const auto & block : blocks)
         dag_writer->write(block);
     dag_writer->flush();
-    dag_writer->finishWrite();
 
     // 4. Start to check write_report.
     size_t per_part_rows = block_rows * block_num / part_num;
@@ -347,7 +344,6 @@ try
     for (const auto & block : blocks)
         dag_writer->write(block);
     dag_writer->flush();
-    dag_writer->finishWrite();
 
     // 4. Start to check write_report.
     size_t expect_rows = block_rows * block_num;

